### DISCUSSION TITLE
No Log and Notifications on Linux Mint 22 "Wilma"

### DISCUSSION THREAD

**Discussion by clriis** [abraunegg/onedrive#2762] [General]
I just did a clean install of Mint 22 on my laptop and then installed 'onedrive'

As Mint 22 is based on Ubuntu 24.04 I am using this repository:

`http://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_24.04  InRelease`

Everything works perfectly, but I don't get a log and notifications are not showing if onedrive is started with 'systemctl --user enable onedrive'

When I check, I get this:

```
claus@TP-Yoga:~$ systemctl status onedrive@claus.service
● onedrive@claus.service - OneDrive Free Client for claus
     Loaded: loaded (/usr/lib/systemd/system/onedrive@.service; enabled; preset: enabled)
     Active: active (running) since Mon 2024-07-29 09:35:43 CEST; 22min ago
       Docs: https://github.com/abraunegg/onedrive
   Main PID: 34962 (onedrive)
      Tasks: 10 (limit: 8959)
     Memory: 32.2M (peak: 35.3M)
        CPU: 4.583s
     CGroup: /system.slice/system-onedrive.slice/onedrive@claus.service
             └─34962 /usr/bin/onedrive --monitor --confdir=/home/claus/.config/onedrive

Jul 29 09:35:43 TP-Yoga onedrive[34962]: Configuration file successfully loaded
Jul 29 09:35:43 TP-Yoga onedrive[34962]: Notification (dbus) server not available, disabling
Jul 29 09:35:43 TP-Yoga onedrive[34962]: Configuring Global Azure AD Endpoints
Jul 29 09:35:46 TP-Yoga onedrive[34962]: Initializing the Synchronization Engine ...
Jul 29 09:35:46 TP-Yoga onedrive[34962]: Initializing monitor ...
Jul 29 09:35:46 TP-Yoga onedrive[34962]: OneDrive monitor interval (seconds): 300
Jul 29 09:35:46 TP-Yoga onedrive[34962]: Starting a sync with OneDrive
Jul 29 09:35:46 TP-Yoga onedrive[34962]: Syncing changes and items from OneDrive ...
Jul 29 09:35:47 TP-Yoga onedrive[34962]: Performing a database consistency and integrity check on locally stored data due to fulls>
Jul 29 09:36:23 TP-Yoga onedrive[34962]: Sync with OneDrive is complete

```

As you can see, the service fails to register with 'dbus'.

Is there a tweak to rectify this?

**Comment by abraunegg** [comment]
@clriis 
Please read the following: https://github.com/abraunegg/onedrive/blob/onedrive-v2.5.0-release-candidate-2/docs/usage.md#onedrive-service-running-as-root-user-via-systemd-arch-ubuntu-debian-opensuse-fedora

Essentially this is a *Ubuntu* problem - the user you are running the client as, does not have the correct XDG variables required to make the dbus server work, which is why notifications get disabled.

If you cant get this working - seek support from Ubuntu and/or Linux Mint developers.

**Comment by abraunegg** [comment]
@clriis 
Further to this, the documentation and client code for v2.5.x has been updated to the following:

![image](https://github.com/user-attachments/assets/43d443cc-0301-408f-9de1-dc3c2742cfcd)

An example of GUI notifications (running via Ubuntu 24.x , not Linux Mint 22 however):

![image](https://github.com/user-attachments/assets/1ebc4a9b-ee83-4c7c-a75b-9e105e672129)

Specifically, v2.5.x will:
* Check if the required runtime environment variables exist:
  * XDG_RUNTIME_DIR
  * DBUS_SESSION_BUS_ADDRESS
* Check that a valid D-Bus message bus daemon is operational

In the event that the required runtime environment variables are missing, the following waning will be printed:
```
WARNING: The required environment variables to enable GUI Notifications are not available; GUI notifications are disabled
```

In the event that the D-Bus message bus daemon is not available, the following warning will be printed:
```
WARNING: D-Bus message bus daemon is not available; GUI notifications are disabled
```

In the event that everything is OK, the following will be printed:
```
D-Bus message bus daemon is available; GUI notifications are now enabled
```

This should help clarify why GUI notifications cannot be enabled, rather than the generic unhelpful message that v2.4.x provides (`Notification (dbus) server not available, disabling`).

**Comment by abraunegg** [comment]
@clriis 
On Linux Mint 22 , default install, logging into the system and running the application with GUI notifications compiled in yields the following:

![image](https://github.com/user-attachments/assets/efcbebe1-efdc-4183-81e0-50bcb1051a2c)

If you are still encountering an issue, please validate your environment against the information above.

**Comment by robertjmcintyre** [comment]
(Thanks for this software, btw!)

After having some critical files go out of sync between OneDrive, my laptop, and my desktop I wanted to get notifications working so that I can see if something goes wrong. I am not a systemd expert (or novice, for that matter), but here's what I found.

I'm running Linux Mint 21.3, and OneDrive Client version v2.5.1.1+np1+1.1, installed via the recommended instructions in the install.md file, and running it as a service via the "systemctl --user method".

I'll spare you the 1.5 hours of reading/troubleshooting and just give you the "what I think is going on TL;DR":

I think that systemd is spinning up the OneDrive service at the same time it's spinning up the dbus service, and dbus takes longer than OneDrive so it looks like it's not running. That's why "systemctl --user restart onedrive" causes it to start working (as noted in this bug report: https://github.com/abraunegg/onedrive/issues/934).

My first attempt to fix (like one commenter in the bug above) was to try using After &  Wants in the onedrive.service file, but that didn't work. I feel I can be forgiven for thinking that "After" means "After it's up and running", but that's not _always_ the case. "After" is satisfied as soon as systemd spins up the separate service process for services that are of type "simple" (which dbus is).

I feel like there's several possible solutions for this:
- Add "After" and "Wants" to the onedrive.service file, and push the dbus team to make that service able to signal systemd that it is up and running
- Update the OneDrive client code to allow for the possibility that just because the conditions for GUI notifications aren't right when they're checked the first time, rather than checking and disabling until the service is restarted
- Add "After" and "Wants", and a pre-start delay (I used 15 seconds for no good reason other than "it worked on my system") to the onedrive.service file

I'm currently using the 3rd option, successfully after reboots, logins/logouts/"systemctl --user restart". I worry about what will happen when a new version of the OneDrive client is released (will it overwrite my onedrive.service file?). I think that the first option is probably overly optimistic. The second option is interesting, but that would be up to @abraunegg.

I feel I should give credit to [this](https://unix.stackexchange.com/questions/696546/after-directive-of-systemd-unit-not-working-as-expected) stackexchange post for basically explaining what was going on and starting me down the dbus service type.

Hope this helps someone.

Here's what my onedrive.service file is now:

```
[Unit]
Description=OneDrive Client for Linux
Documentation=https://github.com/abraunegg/onedrive
After=network-online.target
Wants=network-online.target
After=dbus.service
Wants=dbus.service

[Service]
#Commented out hardenings are disabled because they may not work out of the box on your distribution
#If you know what you are doing please try to enable them.

ProtectSystem=full
#PrivateUsers=true
#PrivateDevices=true
ProtectHostname=true
#ProtectClock=true
ProtectKernelTunables=true
#ProtectKernelModules=true
#ProtectKernelLogs=true
ProtectControlGroups=true
RestrictRealtime=true
ExecStartPre=/bin/sleep 15
ExecStart=/usr/bin/onedrive --monitor
Restart=on-failure
RestartSec=3
#Do not restart the service if a --resync is required which is done via a 126 exit code
RestartPreventExitStatus=126
#Time to wait for the service to stop gracefully before forcefully terminating it
TimeoutStopSec=90

[Install]
WantedBy=default.target
```

**Comment by abraunegg** [comment]
@robertjmcintyre 
> I'm currently using the 3rd option, successfully after reboots, logins/logouts/"systemctl --user restart". I worry about what will happen when a new version of the OneDrive client is released (will it overwrite my onedrive.service file?). 

Yes it would (should) ... 

Added this via https://github.com/abraunegg/onedrive/commit/ba1d225dba3ab195c344d8362a6cdd5e7814fb3e - however the binary path really should be /usr/bin/sleep to be cross platform compatible

**Comment by robertjmcintyre** [comment]
Thanks!

Would adding:

```
After=dbus.service
Wants=dbus.service
```



Be useful? That way, as I understand it, it will "want" dbus if it's available, but not "require"/block it if it's not. I think that would be needed to be as correct as possible.

I think that it should still work as you have it since the onedrive start is delayed, but still...

**Reply by abraunegg** [reply]
Can you test in your system *without* those after|wants ?

I am trying to push v2.5.2 as priority *right now*

**Reply by robertjmcintyre** [reply]
It seems to work fine without the after/wants, at least with the quick testing I've done. Without them it's certainly no worse than it was before this change, and should be better in most cases.

Thanks again, and I'll leave you to v2.5.2...

Cheers!
